### PR TITLE
Uberfire JS extensions support in the Web Jar

### PR DESCRIPTION
**Pull request by ederign**
Improving Uberfire Web Jar support for uberfire runtime plugins (JS API) and perspective JSON generated by Layout editor.

I've also moved uberfire showcase for last project position on build in order to add uberfire extensions for it.

To beginning this review and understand how it works, take a look in uberfire-dynamic plugin project and his generated jar that I've dropped in showcase plugin directory (For the web jar infrastructure).

org.uberfire.backend.server.plugin.PluginsStartup is the starting point of the web jar support/scanning in uberfire.

Next tasks/PR:

Use preferences to configure plugins default dir
Create a Maven API to receive a bunch of files and build a web jar programmatically. (That will be used by perspective editor and david's new stuff)

### CODE REVIEW

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/engine/PluginsWatcher.java [hunk 1]
```diff
@@ -40,119 +40,114 @@
...
 
     volatile boolean active;
     private ExecutorService executor;
-    private GwtRuntimePluginLoader pluginLoader;
+    private PluginsJarProcessor pluginsJarProcessor;
 
     /**
-     * Starts the plugin watcher iff the provided plugin directory exists and
+     * Starts the plugins watcher if the provided plugin directory exists and
      * the watcher hasn't already been started.
```
Comment (csadilek) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/engine/PluginsWatcher.java]: iff was written here on purpose. ;). In general singular class names are preferable as it makes everything easier to read. So, PluginWatcher is better than PluginsWatcher (screwdriver, not screwsdriver, or bike room not bikes room). My recommendation would be to go back to singular names.

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+package org.uberfire.backend.server.plugins.api;
+
+
+/**
+ * A simple marker to keep track of all plugins processors.
+ */
+public interface PluginProcessor {
+
+    /**
+     * Check if this processor should process a given plugin
```
Comment (csadilek) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java]: CheckS and ```@param missing````

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+     */
+    boolean shouldProcess(String pluginFileName);
+
+    /**
+     * Removes all plugin from this processor.
```
Comment (csadilek) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java]: all plugins..

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+     */
+    void removeAll();
+
+    /**
+     * Process a given plugin
```
Comment (csadilek) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java]: Processes a deployed plugin and optionally notifies clients...

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     * Checks if a plugin with the given name is registered.
+     *
+     * @param pluginName the name of the plugin, must not be null.
+     * @return true if registered, otherwise false.
+     */
+    boolean isRegistered(String pluginName);
+
+    /**
+     * Check if a plugin file has the valid extension
```
Comment (csadilek) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java]: Checks.

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+     *
+     * @param fileName fileName of the plugin
```
Comment (csadilek) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java]: fileName  the file name of the plugin

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/RuntimePluginProcessor.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+import java.util.Collection;
+
+/**
+ * A simple marker to keep track of all runtime plugins processors.
+ */
+public interface RuntimePluginProcessor extends PluginProcessor {
+
+    /**
+     * Return all available runtime plugins
```
Comment (csadilek) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/RuntimePluginProcessor.java]: Returns.

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/engine/PluginsJarProcessor.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    /**
+     * Clears the plugins and reloads all currently deployed plugins.
+     * Fires a {@link PluginsReloadedEvent} when done.
+     */
+    public void reload() throws IOException {
+        for (PluginProcessor pluginProcessor : pluginProcessors) {
+            pluginProcessor.removeAll();
+        }
```
Comment (csadilek) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/engine/PluginsJarProcessor.java]: Was about to suggest using a lambda, but this seems to be a bug anyway. As you call removeAll on all processors here and then again in loadPlugins.

File: uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+        for (PluginProcessorExtension type : PluginProcessorExtension.values()) {
+            if (fileName.endsWith(type.getExtension())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    enum PluginProcessorExtension {
```
Comment (csadilek) [inline uberfire-backend/uberfire-backend-server/src/main/java/org/uberfire/backend/server/plugins/api/PluginProcessor.java]: I think this enum could be folded into RuntimePluginType. Then we'd only have one...

File: uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-client/src/main/java/org/uberfire/ext/plugin/client/plugins/RuntimePluginsStartup.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+                } finally {
+                    workbench.removeStartupBlocker(RuntimePluginsStartup.class);
+                }
+
+
+            }
+        }).getRuntimePlugins();
+    }
+
+    private boolean isAPerspectivePlugin(RuntimePlugin p) {
```
Comment (csadilek) [inline uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-client/src/main/java/org/uberfire/ext/plugin/client/plugins/RuntimePluginsStartup.java]: usually we'd just call these methods isPerspectivePlugin and isJSPlugin not isA..

**Comment by ederign** [comment]
@csadilek PR updated.

**Comment by csadilek** [comment]
Great, thanks! All coming together. ;)

**Pull request merged**